gateway/ln-gateway/src/lnd.rs

```diff
@@ -0,0 +1,302 @@
+                        let intercept = SubscribeInterceptHtlcsResponse {
+                            payment_hash: htlc.payment_hash,
+                            incoming_amount_msat: htlc.incoming_amount_msat,
+                            outgoing_amount_msat: htlc.outgoing_amount_msat,
+                            incoming_expiry: htlc.incoming_expiry,
+                            short_channel_id: scid,
+                            intercepted_htlc_id: intercepted_htlc_id.into_inner().to_vec(),
+                        };
+
+                        // Send it to gatewayd for processing
+                        match gwd_tx.send(Ok(intercept)).await {
+                            Ok(_) => {
+                                // Keep a reference to LND sender reference so we can later forward
+                                // outcomes on `complete_htlc` rpc
+                                outcomes.lock().await.insert(
+                                    intercepted_htlc_id,
+                                    (Arc::new(lnd_tx.clone()), htlc.incoming_circuit_key),
+                                );
+
+                                None
+                            }
+                            Err(e) => {
+                                error!("Failed to send HTLC to gatewayd for processing: {:?}", e);
+                                Some(cancel_intercepted_htlc(htlc.incoming_circuit_key))
+                            }
+                        }
+                    };
+
+                if response.is_some() {
+                    // TODO: Consider retrying this if the send fails
+                    let _ = lnd_tx.send(response.unwrap()).await.map_err(|e| {
+                        error!("Failed to send response to LND: {:?}", e);
+                        // The HTLC will timeout and LND will automatically cancel it.
+                        GatewayError::Other(anyhow!("Failed to send response to LND"))
+                    });
+                }
+            }
+
+            info!("HTLC subscription stream ended");
+        })
+        .await;
+
+        Ok(Box::pin(ReceiverStream::new(gwd_rx)))
+    }
+
+    async fn complete_htlc(
+        &self,
+        outcome: CompleteHtlcsRequest,
+    ) -> crate::Result<CompleteHtlcsResponse> {
+        let CompleteHtlcsRequest {
+            action,
+            intercepted_htlc_id,
+        } = outcome;
+
+        let hash = match sha256::Hash::from_slice(&intercepted_htlc_id) {
+            Ok(hash) => hash,
+            Err(e) => {
+                error!("Invalid intercepted_htlc_id: {:?}", e);
+                return Err(GatewayError::LnRpcError(tonic::Status::invalid_argument(
+                    e.to_string(),
+                )));
+            }
+        };
+
+        info!("Completing htlc with reference, {:?}", hash);
+
+        if let Some((outcome, incoming_circuit_key)) = self.outcomes.lock().await.remove(&hash) {
+            let htlca_res = match action {
```

Comment:
nit: would prefer spelling out action, I initially thought the a was a typo :)

Reply:
`htlca_res` isn't a great name